Sherlock/replicator.h

```diff
@@ -0,0 +1,296 @@
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*******************************************************************************/
+
+#ifndef CURRENT_SHERLOCK_REPLICATOR_H
+#define CURRENT_SHERLOCK_REPLICATOR_H
+
+#include <functional>
+#include <string>
+#include <thread>
+
+#include "exceptions.h"
+#include "sherlock.h"
+#include "stream_data.h"
+
+#include "../Blocks/HTTP/api.h"
+#include "../Blocks/SS/ss.h"
+
+#include "../Bricks/sync/scope_owned.h"
+#include "../Bricks/waitable_atomic/waitable_atomic.h"
+
+#include "../TypeSystem/Reflection/types.h"
+
+namespace current {
+namespace sherlock {
+
+template <typename STREAM_ENTRY>
+class SubscribableRemoteStream final {
+ public:
+  using entry_t = STREAM_ENTRY;
+
+  class RemoteStream final {
+   public:
+    RemoteStream(const std::string& url, const std::string& top_level_name, const std::string& namespace_name)
+        : url_(url),
+          schema_(Value<reflection::ReflectedTypeBase>(reflection::Reflector().ReflectType<entry_t>()).type_id,
+                  top_level_name,
+                  namespace_name) {}
+
+    void CheckSchema() const {
+      const auto response = HTTP(GET(url_ + "/schema.simple"));
+      if (static_cast<int>(response.code) == 200) {
+        const auto remote_schema = ParseJSON<SubscribableSherlockSchema>(response.body);
+        if (remote_schema != schema_) {
+          CURRENT_THROW(RemoteStreamInvalidSchemaException());
+        }
+      } else {
+        CURRENT_THROW(RemoteStreamDoesNotRespondException());
+      }
+    }
+
+    const std::string GetURLToSubscribe(uint64_t index) const {
+      return url_ + "?i=" + current::ToString(index);
+    }
+
+    const std::string GetURLToTerminate(const std::string& subscription_id) {
+      return url_ + "?terminate=" + subscription_id;
+    }
+
+   private:
+    const std::string url_;
+    const SubscribableSherlockSchema schema_;
+  };
+
+  template <typename F, typename TYPE_SUBSCRIBED_TO>
+  class RemoteSubscriberThread final : public current::sherlock::SubscriberScope::SubscriberThread {
+   public:
+    RemoteSubscriberThread(ScopeOwned<RemoteStream>& remote_stream,
+                           F& subscriber,
+                           uint64_t start_idx,
+                           std::function<void()> done_callback)
+        : valid_(false),
+          done_callback_(done_callback),
+          external_stop_(false),
+          remote_stream_(remote_stream, [this]() { TerminateSubscription(); }),
+          subscriber_(subscriber),
+          index_(start_idx),
+          thread_([this]() { Thread(); }) {
+      valid_ = true;
+    }
+
+    ~RemoteSubscriberThread() {
+      if (valid_) {
+        if (!subscriber_thread_done_) {
```

Comment:
It is declared as `std::atomic_bool`. And what the point of acquiring the mutex here and below? It doesn't prevent us from skipping `subscriber_thread_done_` setting (and it's not bad). By the way, there is exact the same code in `sherlock::Stream::~SubscriberThreadInstance`. But I'm thinking in this case we can remove this check at all, because there is one more inside `TerminateSubscription()`. What do you think?